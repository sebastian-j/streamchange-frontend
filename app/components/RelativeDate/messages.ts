/*
 * RelativeDate Messages
 *
 * This contains all the text for the RelativeDate component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.components.RelativeDate';

export default defineMessages({
  beforeYesterday: {
    id: `${scope}.beforeYesterday`,
    defaultMessage: 'Two days ago',
  },
  fewMinutesAgo: {
    id: `${scope}.fewMinutesAgo`,
    defaultMessage: '{value} minutes ago',
  },
  justNow: {
    id: `${scope}.justNow`,
    defaultMessage: 'Just now',
  },
  minuteAgo: {
    id: `${scope}.minuteAgo`,
    defaultMessage: 'Minute ago',
  },
  minutesAgo: {
    id: `${scope}.minutesAgo`,
    defaultMessage: '{value} minutes ago',
  },
  never: {
    id: `${scope}.never`,
    defaultMessage: 'Never',
  },
  secondsAgo: {
    id: `${scope}.secondsAgo`,
    defaultMessage: '{value} seconds ago',
  },
  today: {
    id: `${scope}.today`,
    defaultMessage: 'Today',
  },
  yesterday: {
    id: `${scope}.yesterday`,
    defaultMessage: 'Yesterday',
  },
});
