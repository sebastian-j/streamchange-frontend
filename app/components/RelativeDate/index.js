import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, FormattedDate } from 'react-intl';
import messages from './messages';

function RelativeDate(props) {
  const dt = new Date(props.ISO8601Date);
  const now = new Date();
  let convertedDate = '';
  if (
    dt.getFullYear() === now.getFullYear() &&
    dt.getMonth() === now.getMonth() &&
    dt.getDate() + 2 === now.getDate()
  ) {
    return (
      <span className={props.className}>
        <FormattedMessage {...messages.beforeYesterday} />
        {` ${dt.getHours()}:${
          dt.getMinutes() < 10 ? '0' : ''
        }${dt.getMinutes()}`}
      </span>
    );
  }
  if (
    dt.getFullYear() === now.getFullYear() &&
    dt.getMonth() === now.getMonth() &&
    dt.getDate() + 1 === now.getDate()
  ) {
    return (
      <span className={props.className}>
        <FormattedMessage {...messages.yesterday} />
        {` ${dt.getHours()}:${
          dt.getMinutes() < 10 ? '0' : ''
        }${dt.getMinutes()}`}
      </span>
    );
  }
  if (now.getTime() - dt.getTime() < 10000) {
    return (
      <span className={props.className}>
        <FormattedMessage {...messages.justNow} />
      </span>
    );
  }
  if (now.getTime() - dt.getTime() < 60000) {
    return (
      <span className={props.className}>
        <FormattedMessage
          {...messages.secondsAgo}
          values={{
            value: `${Math.round((now.getTime() - dt.getTime()) / 1000)}`,
          }}
        />
      </span>
    );
  }
  if (now.getTime() - dt.getTime() < 120000) {
    return (
      <span className={props.className}>
        <FormattedMessage {...messages.minuteAgo} />
      </span>
    );
  }
  if (now.getTime() - dt.getTime() < 270000) {
    return (
      <span className={props.className}>
        <FormattedMessage
          {...messages.fewMinutesAgo}
          values={{
            value: `${Math.round((now.getTime() - dt.getTime()) / 60000)}`,
          }}
        />
      </span>
    );
  }
  if (now.getTime() - dt.getTime() < 3600000) {
    return (
      <span className={props.className}>
        <FormattedMessage
          {...messages.minutesAgo}
          values={{
            value: `${Math.round((now.getTime() - dt.getTime()) / 60000)}`,
          }}
        />
      </span>
    );
  }
  if (
    dt.getFullYear() === now.getFullYear() &&
    dt.getMonth() === now.getMonth() &&
    dt.getDate() === now.getDate()
  ) {
    return (
      <span className={props.className}>
        <FormattedMessage {...messages.today} />
        {` ${dt.getHours()}:${
          dt.getMinutes() < 10 ? '0' : ''
        }${dt.getMinutes()}`}
      </span>
    );
  }
  if (props.ISO8601Date === null) {
    return (
      <span className={props.className}>
        <FormattedMessage {...messages.never} />
      </span>
    );
  }
  convertedDate = ` ${dt.getHours()}:${
    dt.getMinutes() < 10 ? '0' : ''
  }${dt.getMinutes()}`;
  return (
    <span className={props.className}>
      <FormattedDate
        value={new Date(dt)}
        year="numeric"
        month="long"
        day="2-digit"
      />
      {convertedDate}
    </span>
  );
}

RelativeDate.propTypes = {
  className: PropTypes.string,
  ISO8601Date: PropTypes.string.isRequired,
};

RelativeDate.defaultProps = {
  className: 'relativeDate',
};

export default RelativeDate;
