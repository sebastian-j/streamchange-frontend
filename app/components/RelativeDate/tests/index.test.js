import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import RelativeDate from '../index';

const renderer = new ShallowRenderer();

describe('<RelativeDate />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(
      <RelativeDate ISO8601Date="2019-12-24T07:27:56.27-00:00" />,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
