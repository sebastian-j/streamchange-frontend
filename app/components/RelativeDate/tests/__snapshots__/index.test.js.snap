// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RelativeDate /> should render and match the snapshot 1`] = `
<span
  className="relativeDate"
>
  <FormattedDate
    day="2-digit"
    month="long"
    value={2019-12-24T07:27:56.270Z}
    year="numeric"
  />
   8:27
</span>
`;
